source/commoncrypto/commoncrypto_ecc.c

```diff
@@ -0,0 +1,607 @@
+    enum aws_ecc_curve_name curve_name;
+    if (aws_ecc_curve_name_from_oid(&oid, &curve_name)) {
+        goto error;
+    }
+
+    size_t key_coordinate_size = s_key_coordinate_byte_size_from_curve_name(curve_name);
+
+    struct aws_byte_cursor *private_key = NULL;
+    struct aws_byte_cursor *public_key = NULL;
+
+    size_t public_key_blob_size = key_coordinate_size * 2 + 1;
+
+    if (pair_part_1.ptr && pair_part_1.len) {
+        if (pair_part_1.len == key_coordinate_size) {
+            private_key = &pair_part_1;
+        } else if (pair_part_1.len == public_key_blob_size) {
+            public_key = &pair_part_1;
+        }
+    }
+
+    if (pair_part_2.ptr && pair_part_2.len) {
+        if (pair_part_2.len == key_coordinate_size) {
+            private_key = &pair_part_2;
+        } else if (pair_part_2.len == public_key_blob_size) {
+            public_key = &pair_part_2;
+        }
+    }
+
+    if (!private_key && !public_key) {
+        aws_raise_error(AWS_ERROR_CAL_MISSING_REQUIRED_KEY_COMPONENT);
+        goto error;
+    }
+
+    struct aws_byte_cursor pub_x_cur;
+    struct aws_byte_cursor pub_y_cur;
+    struct aws_byte_cursor *pub_x = NULL;
+    struct aws_byte_cursor *pub_y = NULL;
+
+    if (public_key) {
+        aws_byte_cursor_advance(public_key, 1);
+        pub_x_cur = *public_key;
+        pub_x_cur.len = key_coordinate_size;
+        pub_y_cur.ptr = public_key->ptr + key_coordinate_size;
+        pub_y_cur.len = key_coordinate_size;
+        pub_x = &pub_x_cur;
+        pub_y = &pub_y_cur;
+    }
+
+    struct commoncrypto_ecc_key_pair *cc_key_pair =
+        s_alloc_pair_and_init_buffers(allocator, curve_name, pub_x, pub_y, private_key);
+
+    CFDataRef key_data = CFDataCreate(
+        cc_key_pair->cf_allocator, cc_key_pair->key_pair.key_buf.buffer, cc_key_pair->key_pair.key_buf.len);
+
+    CFMutableDictionaryRef key_attributes = CFDictionaryCreateMutable(cc_key_pair->cf_allocator, 6, NULL, NULL);
+    CFDictionaryAddValue(key_attributes, kSecAttrKeyType, kSecAttrKeyTypeECSECPrimeRandom);
+
+    if (private_key) {
+        CFDictionaryAddValue(key_attributes, kSecAttrKeyClass, kSecAttrKeyClassPrivate);
+
+        CFDictionaryAddValue(key_attributes, kSecAttrCanSign, kCFBooleanTrue);
+        CFDictionaryAddValue(key_attributes, kSecAttrCanDerive, kCFBooleanTrue);
+
+        if (public_key) {
+            CFDictionaryAddValue(key_attributes, kSecAttrCanVerify, kCFBooleanTrue);
+        }
+    } else if (public_key) {
+        CFDictionaryAddValue(key_attributes, kSecAttrKeyClass, kSecAttrKeyClassPublic);
+        CFDictionaryAddValue(key_attributes, kSecAttrCanSign, kCFBooleanFalse);
+        CFDictionaryAddValue(key_attributes, kSecAttrCanVerify, kCFBooleanTrue);
+    }
+
+    CFErrorRef error = NULL;
+
+    cc_key_pair->priv_key_ref = SecKeyCreateWithData(key_data, key_attributes, &error);
+
+    if (public_key) {
+        cc_key_pair->pub_key_ref = SecKeyCopyPublicKey(cc_key_pair->priv_key_ref);
```

Comment:
not checking result of SecKeyCopyPublicKey()

Reply:
In this case it doesn't matter, because we don't dereference it and the key is still valid, it just can't validate signatures.